### PR TITLE
[ConstraintGraph] Disallow subtype constraint contractions

### PR DESCRIPTION
**Pull request by xedin**
Presently subtype constraint is considered a candidate for contraction
via `shouldContractEdge` when left-hand side of the subtype constraint
is an inout type with type variable inside. Although it's intended to
be used only in combination with BindParam constraint assigned to the
same variable, that is actually never checked and contraction of subtype
constraint like that is invalid.

Resolves: rdar://problem/34333874

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please smoke test

File: validation-test/Sema/rdar34333874.swift [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+   try update(&this)
+   return this
+}
+
+let rdar34333874 = foo(C()) {
+  $0.a = 42
+  $0.b = 0
+}
+
+print(rdar34333874)
```
Comment (rudkx) [inline validation-test/Sema/rdar34333874.swift]: newline

**Comment by xedin** [comment]
@swift-ci please smoke test

File: validation-test/Sema/rdar34333874.swift [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+func foo<T>(_ item: T, update: (inout T) throws -> Void) rethrows -> T {
+   var this = item
+   try update(&this)
+   return this
+}
+
+let rdar34333874 = foo(C()) {
+  $0.a = 42
```
Comment (rudkx) [inline validation-test/Sema/rdar34333874.swift]: Is this really testing the same thing as the reported bug? This actually is mutating the input whereas the bug report wasn't.

**Review by rudkx — COMMENTED** [review]
Otherwise LGTM.

File: validation-test/Sema/rdar34333874.swift [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// RUN: rm -rf %t
+// RUN: mkdir -p %t
+// RUN: %target-build-swift %s -o %t/a.out
+// RUN: %target-run %t/a.out
```
Comment (jrose-apple) [inline validation-test/Sema/rdar34333874.swift]: Drive-by comment: This should be `%target-run-simple-swift`, which gets run in more situations.

File: validation-test/Sema/rdar34333874.swift (hunk 2 above)
Comment (xedin) [reply validation-test/Sema/rdar34333874.swift]: Example from the rdar does modify fields of `$0`

Comment (swiftix) [reply validation-test/Sema/rdar34333874.swift]: The actual original code did mutate input inside the closure. I'd suggest testing both cases to be on the safe side.

File: validation-test/Sema/rdar34333874.swift (hunk 3 above)
Comment (xedin) [reply validation-test/Sema/rdar34333874.swift]: Thanks, @jrose-apple ! I didn't know that existed...

File: validation-test/Sema/rdar34333874.swift (hunk 2 above)
Comment (xedin) [reply validation-test/Sema/rdar34333874.swift]: Sure, sounds good, I'll add one which doesn't mutate the fields.

**Comment by xedin** [comment]
@swift-ci please smoke test

File: validation-test/Sema/rdar34333874.swift [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+@inline(never)
+func foo<T>(_ item: T, update: (inout T) throws -> Void) rethrows -> T {
+  var this = item
+  try update(&this)
+  return this
+}
+
+let rdar34333874 = foo(C()) {
```
Comment (rudkx) [inline validation-test/Sema/rdar34333874.swift]: Hmm, apparently I didn't submit my comment about the test. Are you sure this is testing the same thing as the original report? This appears to actually be mutating something whereas the original report wasn't.

Comment (xedin) [reply validation-test/Sema/rdar34333874.swift]: Both me and @swiftix replied to that one and I've added one more test in addition to the mutating one from SR-5861 which does mutate the fields of the argument.

**Comment by xedin** [comment]
I've fixed a problem in `isStrictInoutSubtypeConstraint` which returned true only if the left-hand side of the subtype with inout is not a type-variable, which allowed bind params to contracted incorrectly and added one more test with a single statement closure.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility
